From 5cca2fa5997292a87302bdc7e7ed3231371e98bd Mon Sep 17 00:00:00 2001
From: Alex Kube <alexander.j.kube@gmail.com>
Date: Wed, 23 Oct 2019 21:15:37 +0430
Subject: [PATCH 2/9] cmd/go: Allow GOTOOLDIR to be overridden in the
 environment

to allow for split host/target build roots

Adapted to Go 1.13 from patches originally submitted to
the meta/recipes-devtools/go tree by
Matt Madison <matt@madison.systems>.

Upstream-Status: Inappropriate [OE specific]

Signed-off-by: Alexander J Kube <alexander.j.kube@gmail.com>
---
 src/cmd/dist/build.go          | 4 +++-
 src/cmd/go/internal/cfg/cfg.go | 6 +++++-
 2 files changed, 8 insertions(+), 2 deletions(-)

--- a/src/cmd/dist/build.go
+++ b/src/cmd/dist/build.go
@@ -259,7 +259,9 @@ func xinit() {
 	}
 	xatexit(rmworkdir)
 
-	tooldir = pathf("%s/pkg/tool/%s_%s", goroot, gohostos, gohostarch)
+	if tooldir = os.Getenv("GOTOOLDIR"); tooldir == "" {
+		tooldir = pathf("%s/pkg/tool/%s_%s", goroot, gohostos, gohostarch)
+	}
 
 	goversion := findgoversion()
 	isRelease = strings.HasPrefix(goversion, "release.") || strings.HasPrefix(goversion, "go")
--- a/src/cmd/go/internal/cfg/cfg.go
+++ b/src/cmd/go/internal/cfg/cfg.go
@@ -231,7 +231,11 @@ func SetGOROOT(goroot string, isTestGo b
 			// This matches the initialization of ToolDir in go/build, except for
 			// using ctxt.GOROOT and the installed GOOS and GOARCH rather than the
 			// GOROOT, GOOS, and GOARCH reported by the runtime package.
-			build.ToolDir = filepath.Join(GOROOTpkg, "tool", installedGOOS+"_"+installedGOARCH)
+	   		if s := os.Getenv("GOTOOLDIR"); s != "" {
+				build.ToolDir = filepath.Clean(s)
+			} else {
+				build.ToolDir = filepath.Join(GOROOTpkg, "tool", installedGOOS+"_"+installedGOARCH)
+			}
 		}
 	}
 }
